src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
   const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
-  const [isOpen, setIsOpen] = React.useState(false);
+  const optionsRef = React.useRef<Array<HTMLLIElement>>([]);
+
+  // Latest value of the onChange prop
+  const onChangeProp = React.useRef(onChange);
+  React.useEffect(() => {
+    onChangeProp.current = onChange;
+  }, [onChange]);
+
+  // Reducer where all the component-wide state is stored
+  const [{ isOpen, selectedIndex, focusedIndex }, dispatch] = React.useReducer(
+    comboBoxReducer,
+    {
+      isOpen: false,
+      selectedIndex: -1,
+      focusedIndex: -1,
+    },
+  );
+
+  React.useEffect(() => {
+    if (isOpen) {
+      // Make sure the input is focused when the dropdown opens
+      inputRef.current?.focus();
+    } else {
+      // Reset the focused index when the dropdown closes
+      dispatch(['focus']);
+    }
+  }, [isOpen]);
 
   // Set min-width of menu to be same as input
-  const [minWidth, setMinWidth] = React.useState(0);
+  const [minWidth, setMinWidth] = React.useState(300);
   React.useEffect(() => {
     if (inputRef.current) {
       setMinWidth(inputRef.current.offsetWidth);
     }
   }, [isOpen]);
 
+  // Initialize filtered options to the latest value options
   const [filteredOptions, setFilteredOptions] = React.useState(options);
   React.useEffect(() => {
     setFilteredOptions(options);
   }, [options]);
 
-  const [focusedIndex, setFocusedIndex] = React.useState(() =>
-    options.findIndex((option) => value === option.value),
-  );
-
-  // Maintain internal selected value state synced with `value` prop
-  const [selectedValue, setSelectedValue] = React.useState<T | undefined>(
-    value,
-  );
-  React.useEffect(() => {
-    setSelectedValue(value);
-  }, [value]);
-
-  // Controlled input value
-  const [inputValue, setInputValue] = React.useState<string>('');
-  const onInput = React.useCallback(
+  // Filter options based on input value
+  const [inputValue, setInputValue] = React.useState(inputProps?.value ?? '');
+  const handleOnInput = React.useCallback(
     (event: React.ChangeEvent<HTMLInputElement>) => {
-      setInputValue(event.target.value);
+      const { value } = event.currentTarget;
+      setInputValue(value);
+      dispatch(['open']);
+      setFilteredOptions(
+        filterFunction?.(options, value) ??
+          options.filter((option) =>
+            option.label.toLowerCase().includes(value.toLowerCase()),
+          ),
+      );
       inputProps?.onChange?.(event);
     },
-    [inputProps],
+    [filterFunction, inputProps, options],
   );
 
-  // update inputValue and focusedIndex every time selected value changes
+  // Reset focused item when filter changes
   React.useEffect(() => {
-    const selectedOption = options.find(({ value }) => value === selectedValue);
-    setInputValue(selectedOption?.label ?? '');
-    setFocusedIndex(selectedOption ? options.indexOf(selectedOption) : -1);
-  }, [selectedValue, options]);
+    dispatch(['focus']);
```

Comment:
Can you move this call to `handleInput`?

Reply:
Why? `filteredOptions` could also change if `optionsProp` changes for some reason